scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketTransform.java

```diff
@@ -342,104 +347,237 @@ public int hashCode() {
+                final Iterable<?> maybeUnfinishedIt =
+                    mergedKeyGroup.getValue().getAll(source.getTupleTag());
+                if (SortedBucketSource.TraversableOnceIterable.class.isAssignableFrom(
+                    maybeUnfinishedIt.getClass())) {
+                  ((SortedBucketSource.TraversableOnceIterable<?>) maybeUnfinishedIt)
+                      .ensureExhausted();
+                }
+              });
+        } catch (Exception ex) {
+          throw new RuntimeException("Failed to write merged key group", ex);
         }
-      };
+      }
+      outputCollector.onComplete();
+      out.output(new MergedBucket(bucketId, dst, effectiveParallelism));
+    }
+  }
+
+  private static class NoSides<FinalKeyT, FinalValueT>
+      extends TransformDoFn<FinalKeyT, FinalValueT> {
+    private final TransformFn<FinalKeyT, FinalValueT> transformFn;
+
+    public NoSides(
+        List<SortedBucketSource.BucketedInput<?, ?>> sources,
+        SourceSpec<FinalKeyT> sourceSpec,
+        SMBFilenamePolicy.FileAssignment fileAssignment,
+        FileOperations<FinalValueT> fileOperations,
+        TransformFn<FinalKeyT, FinalValueT> transformFn,
+        Distribution keyGroupSize) {
+      super(sources, sourceSpec, fileAssignment, fileOperations, keyGroupSize);
+      this.transformFn = transformFn;
+    }
+
+    @Override
+    protected void outputTransform(
+        final KV<FinalKeyT, CoGbkResult> mergedKeyGroup,
+        final ProcessContext context,
+        final OutputCollector<FinalValueT> outputCollector,
+        final BoundedWindow window) {
+      transformFn.writeTransform(mergedKeyGroup, outputCollector);
+    }
+  }
+
+  private static class WithSides<FinalKeyT, FinalValueT>
+      extends TransformDoFn<FinalKeyT, FinalValueT> {
+    private final TransformFnWithSideInputContext<FinalKeyT, FinalValueT> transformFn;
+
+    public WithSides(
+        List<SortedBucketSource.BucketedInput<?, ?>> sources,
+        SourceSpec<FinalKeyT> sourceSpec,
+        SMBFilenamePolicy.FileAssignment fileAssignment,
+        FileOperations<FinalValueT> fileOperations,
+        TransformFnWithSideInputContext<FinalKeyT, FinalValueT> transformFn,
+        Distribution keyGroupSize) {
+      super(sources, sourceSpec, fileAssignment, fileOperations, keyGroupSize);
+      this.transformFn = transformFn;
+    }
+
+    @Override
+    protected void outputTransform(
+        final KV<FinalKeyT, CoGbkResult> mergedKeyGroup,
+        final ProcessContext context,
+        final OutputCollector<FinalValueT> outputCollector,
+        final BoundedWindow window) {
+      transformFn.writeTransform(mergedKeyGroup, context, outputCollector, window);
+    }
+  }
+
+  private static class BucketReader extends BoundedSource.BoundedReader<BucketItem> {
+    private final BoundedSource<BucketItem> currentSource;
+    private boolean started;
+    // there is only ever a single item in this source
+    @Nullable private BucketItem next;
```

Comment:
Maybe not worth having `BucketItem` be its own class since it's only used once

Reply:
IIRC I did this to get the types to be clear e.g. here: 
https://github.com/spotify/scio/pull/3915/files#diff-750c44e50ebc665919ed39fdf777fe6a3607e6ff25386bbfd874eed26956921eR536